yarl/_url.py

```diff
@@ -80,6 +77,15 @@
     Self = Any
 
 
+class UndefinedType(Enum):
+    """Singleton type for use with not set sentinel values."""
+
+    _singleton = 0
+
+
+UNDEFINED = UndefinedType._singleton
```

Comment:
Surprised to see the whole enum machinery here rather than `UNDEFINED = object()` but I suppose the costs are only once at import time

Reply:
The typing gets messy otherwise as any object would be allowed